File: blind 75/Dynamic Programming and Recursion/longest_increasing_subsequence.py
'''
https://leetcode.com/problems/longest-increasing-subsequence/solution/
DP Solution
https://www.youtube.com/watch?v=CE2b_-XfVDk&t=324s&ab_channel=TusharRoy-CodingMadeSimple

Top Down Solution
https://www.youtube.com/watch?v=ekKYRYFEm9w&list=PLQdWvigIOnscz0Fgps9PtnymVkvJrZylH&index=5&ab_channel=AnuragVishwa
'''

"""
Revisited on 8/21/2023
adapted from: 
https://leetcode.com/problems/longest-increasing-subsequence/discuss/1124603/Top-DownBackTrackingMemoization
Top down approach:
1) Find increasingly larger numbers within the array
2) when one is found, make a recursive call to find the next greatest number, where the result of this
recursive call will be the max length subsequence that can be found starting at this number. 
Pass in the index at which this element was found.
3) For memoization, we store the max length mapped to each index of nums.
4) Our "base case" is technically when we pass in an element, and no greater number is found.
In that case, the length of the subsequence is just 1, so we return the length

In the leetcode problem, it passes in the index instead of a smaller slice of the array which is smart
to save space. That way, we can just key into the original nums array and know the index which makes
memoization much easier.

Time complexity: O(N^2) (one loop per element, and then additional recursive calls from start to n-1)
Space Complexity: O(N^2) recursive calls

Example:
nums = [10, 9, 2, 5, 3, 7, 101, 18]

1) 
self.memo = {}
first call:
starting from i = 0
10

loops through 9 ... 18
finds 101 > 10

num = 101
start = 6
dfs(6, 101)

cannot find a number greater than 101,
therefore at index 6, the greatest length we get is 1

self.memo = {6: 1}

2) 
Back to 10, we store in our memo index 0 mapped to length 2, exit and now run dfs starting from i = 1 (9)
9

loops through 2 ... 18
finds 101
we already found 101 in our memoization, since index 6 is in memo
return 1

therefore, the max is now 1 + self.memo[6] which is 2
store memo, which is {1: 2} 

goes back to loop, finds 18

at 18, max length is 1 since there's no other elements past it
stores index 7 mapped to length 1 in self.memo, {7: 1}

self.memo = {0: 2, 6:1, 7:1, 1: 2}

3) 
exits, goes back to start running DFS on i = 2 now (element 2)
it will now loop through
5 ... 18
finds 5, does a recursive call

loops through 3 ... 18
finds 7, does a recursive call

loops through 101 ... 18
finds 101, this is already in our memoization so returns 1
1+1 = 2 (current subsequence is 7, 101)

finds 18, this is also already in our memoization so returns 1
1+1 = 2 

We store length 3 for the index 4
self.memo = {0: 2, 6:1, 7:1, 1:2, 4: 3}

goes back to 5 ... 18, adds one to the length
we store length 4 for the index2 
self.memo = {0: 2, 6:1, 7:1, 1:2, 4: 3, 2: 4}

This same process continues until all i ... n - 1 are examined,
the final answer should be 4, with the longest subsequence here
being 2, 3, 7, 101 OR 2, 5, 7, 101


"""
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        """
        Revisited on 1/25/2024
        O(N^2) Time Complexity (Same as above)
        1) Approach, find all possibilities, pick the maximum length at each subproblem
        10 9 2 5 3 7 101 18
        2 3 7 101
        
        10 101
        9 101
        
        2 3 OR 2 7
        Note that at 2, we have a choice of picking either 3 as the next element, OR
        7 as the next element, these are both valid increasing subsequences,
        so this is the first instance of the recursive sub problem. You can see 
        the decision tree that you'd have to make for each subsequent increasing number.
        
                           2 3 OR 2 7
                           
            2 3 7 OR 2 3 101           2 7 101 OR 2 7 18
                     
         2 3 7 18 OR 2 3 7 101                       
         
        
        
        """
        # we can store at a given index,
        # what's the longest increasing subsequence we can make at that index?
        # this will cut down on repeated subproblems
        self.memo = dict()
        def search(i):
            maxLen = 1
            if i in self.memo:
                return self.memo[i]
            for k in range(i, len(nums)):
                # if we find a number greater than our 
                # current number that also has a greater index,
                # this is a subproblem, because there's a chance
                # we could find multiple numbers greater than our current number,
                # which would be different subsequences, so do a recursive call.
                if nums[k] > nums[i]:
                    maxLen = max(1 + search(k), maxLen)
            self.memo[i] = maxLen
            return self.memo[i]
        
        maxLen = 1
        # for each i, we want to find a number greater than nums[i]
        for i in range(len(nums)):
            maxLen = max(search(i), maxLen)
        return maxLen


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        self.memo = dict()
        self.nums = nums
        def dfs(start, num):
            if start in self.memo:
                return self.memo[start]
            length = 1
            for i in range(start, len(self.nums)):
                if self.nums[i] > num:  
                    length = max(length, 1 + dfs(i+1, self.nums[i]))            
            self.memo[start] = length
            return length
        maxLength = 0
        for i in range(len(self.nums)):
            maxLength = max(dfs(i+1, nums[i]), maxLength)
        return maxLength

class Solution:
    ## converted from Tushar Roy's java solution
    ## https://github.com/mission-peace/interview/blob/master/src/com/interview/dynamic/LongestIncreasingSubsequence.java
    def lengthOfLIS(self, nums: List[int]) -> int:
        ## the longest increasing subsequence of the smallest subproblem (which is just the individual number itself) is always 1
        ## so we initialize the position to 1
        dp = [1] * len(nums)
        ## we skip the first element since the longest subsequence of the first element is always 1
        for i in range(1, len(nums)):
            for j in range(i):
            	## the recurrence relation is that if we encounter nums[i],if nums[i] > nums[j], we check the dp[i] position
     			## as well as the dp[j] position + 1 and pick whichever one is greater
     			## the reason why is that dp[i] will always contain the longest increasing subsequence at that point
     			## so we can use each previous step and add one to get the answer
                if (nums[i] > nums[j]):
                    dp[i] = max(dp[i], dp[j] + 1)
        ## by the end, the max of dp array will contain the longest increasing subsequence at that point
        return max(dp)


    ## recursive solution using brute force O(2^N)
    ## https://stackoverflow.com/questions/37561909/does-there-exist-a-top-down-dynamic-programming-solution-for-longest-increasing
    ## https://www.youtube.com/watch?v=ekKYRYFEm9w&list=PLQdWvigIOnscz0Fgps9PtnymVkvJrZylH&index=5&ab_channel=AnuragVishwa
    def lengthOfLIS(self, nums: List[int]) -> int:
    	return self.search(nums, float('-inf'), 0)

    ## the idea mentioned in the youtube video is that at each index
    ## we have the decision of either including the current item in our current subsequence and continue onto the next item
    ## OR we can exclude the current item and continue onto the next item
    ## and we'll recursively repeat this (either including or excluding the current item) and figure out the max between these two decision
    ## to determine the longest increasing subsequence
    def search(self, nums: List[int], prev, current) -> int:
    	## base case, if our current index is the length of the array, return 0
    	if (current == len(nums)):
    		return 0
    	include = 0	
		## if our current element is greater than our previous element we can make the decision to include it in our
		## subsequence
    	if (nums[current] > prev):
    		## pass in the current element down into prev, and increase the current index
    		include = 1 + self.search(nums, nums[current], current+1)
    	## in another recursive call, we'll figure out the total amount if we exclude the current element
    	exclude = self.search(nums, prev, current+1)
    	## then we take the max between the two to figure out which total was larger
    	return max(include, exclude)

 
            
        
        
        
               
                
        

File: node_js_ds_and_algos_implementations/1D dynamic programming/longest_increasing_subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
/*
Approach:
since we're only concerned with the length of the longest increasing subsequence, rather than the actual subsequence itself, we
keep track of a length, and when we see a number that is greater than previous, we increase the length. There's also the option to not 
count this in the subsequence, in that case, we'd just continue incrementing

Base Case: if we've reached the end of the list, the subsequence must've been at least a length of 1,
so return 1

With memoization:
O(N^2) time solution, since we still need to run an O(N) recursive search
on each number in the list. The memoization saves time within the recursive search
by cutting the time to O(N) instead of exponential 

Space: O(N)

Example:
[10, 14, 11, 12]

1st call:
start = 0
end = 0
nums[end] is not greater than nums[start]
length2 = search(start, end+1)

2nd call
start = 0
end = 1
nums[end] > nums[start]
len = search(end, end+1) + 1

3rd call
start = 1
end = 2
nums[2] is not greater than nums[1]
length2 = search(start, end+1)

4th call
start = 1
end = 3
nums[3] is not greater than nums[1]

5th call
start = 1
end = 4

base case reached, 4 >= N, returns 1

Back to the 4th call
start = 1
end = 3
Math.max(1, 1) = 1
returns 1

Back to the 3rd call
start = 1
end = 2
Math.max(1, 1) = 1
returns 1

Back to the 2nd call
start = 0
end = 1
len = 1 + 1 = 2
try the second route

length2 = search(start, end+1)

6th call
start = 0
end = 2
nums[2] > nums[0], 
len = search(end, end+1)+1

7th call
start = 2
end = 3
nums[3] > nums[2]
len = search(end, end+1)+1

8th call
start = 3
end = 4
base case reached (4 >= N)
returns 1

back to 7th call
start = 2
end = 3
len = 1 + 1 = 2

back to 6th call
start = 0
end = 2
len = 2 + 1 = 3

try 2nd route
length2 = search(start, end+1)

7th call
start = 0
end = 3
nums[3] > nums[0]
len = search(end, end+1)+1

8th call
start = 0
end = 4
base case reached (4 >= N), returns 1

back to 7th call
start = 0
end = 3
len = 1 + 1 = 2

it'll try 0 and 4, but won't work and will return 1, at this point
Math.max(2, 1) = 2, so return 2

back to 6th call
start = 0
end = 2
len = 3
length2 = 2
Math.max(3, 2) = 3

back to 2nd call
length2 = 3
len = 2 
Math.max(2, 3) = 3

back to 1st call
Math.max(1, 3) = 3

returns 3 here

In the for loop, we now try the starting index at 1
this is where the memo object comes in, since we've already 
calculated the max that can be reached at 1 (which is 1), we can simply just return 1

same with the other indices i = 2, i = 3, their answers are 2 and 1 respectively,
none of which are larger than 3

For another case like though
[11, 9, 10, 12]
we'd need this for loop in order to consider 9 as a starting point, otherwise,
since we start at 11, we'd skip 9 and 10 and consider 12 (for a length of 2)

in the loop when i = 1,
If we try starting from 9 here, we'd calculate 3 which is our final answer.

*/
var lengthOfLIS = function(nums) {
    let N = nums.length
    let memo = {}
    var search = function(start, end, i) {
        // base case, if we've reached the end, the length of the subsequence was at least 1, so return 1
        if (end >= N){
            return 1
        }
        if (start in memo){
            return memo[start]
        }
        // if we find an increasing number, set the 
        // new start to be end
        let len = 1
        if (nums[end] > nums[start]){
            len = search(end, end+1)+1
        }
        // we also have the option of not choosing this number
        // to include in our subsequence and looking for a different increasing number, in this case we keep start the same, and increment end
        // we also want to find the max length between if we were to choose
        // the increasing element, or if we were to ignore it and find another one
        let length2 = search(start, end+1)
        let maxLength = Math.max(len, length2)
        // at this start index, store the length of the longest
        // increasing subsequence starting at this index
        memo[start] = maxLength
        return maxLength
    }
    let res = Number.NEGATIVE_INFINITY
    for (let i = 0; i < N; ++i){
        res = Math.max(res, search(i, i))
    }
    return res
};